### DISCUSSION TITLE
Proposal to drop stored cookies and replace them with a deterministic hash

### DISCUSSION THREAD

**Discussion by nagy** [seaweedfs/seaweedfs#2133] [Ideas]
# Motivation
Save 4 bytes per file in the Volume storage.

# My Understanding
For each file, the cookie serves to prevent rogue clients from iterating over
all files on the volume server. It is, for each file, generated as a random
number consuming 4 bytes, and those 4 bytes are added into the volumes `.dat`
files.

# Proposal
Introduce a per-volume seed with enough entropy ( lets say 32 bytes ) and store
it next to the volume. This could be in the superblock header or in a `.seed`
file next to the volume. This seed should be kept hidden/private from clients.

Then, upon creation of a new file take the `Hash( <fileid> | <seed> )` where `Hash`
could be the SHA3 function. Note that `|` denotes a concatenation. Cut it to 4
bytes. This is the new cookie.

Upon a file request, the Hash is calculated again and matched against the
request from the client.

# Fun fact
This is very similar to how SYN-Cookies work within TCP.
https://en.wikipedia.org/wiki/SYN_cookies

**Comment by prologic** [comment]
The only downside I can say is that you were trading storage space for computational workload per request? 🤔￼

**Comment by chrislusf** [comment]
what is the benefit?

**Reply by nagy** [reply]
Smaller overhead per file in the volume storage. Right now it is 40, and with this it could be 36.

**Reply by chrislusf** [reply]
One of the cons is to change the workflow and the file id is only finalized when the volume stored the file.
